### ISSUE TITLE
Disable home-row mod keys when typing quickly

### ISSUE THREAD

**Issue by MaxGyver83** [david-janssen/kmonad#228]
## Feature request

The `tap-hold-next-release` keys make it easy to have modifier keys in the home row. It works as expected most of the time. But after some months, I still make errors:

1. I want to press a shortcut but I get two letters (because I release the first key too quickly).
2. I'm writing text and suddently I fire off a shortcut (which deletes the text I have written in the worst case).

Number 1 is not a big deal. If this happen now and then, I can easily correct that mistake.
But number 2 is really annoying and it could be prevented by KMonad. When I type text quickly (for example, x keys in a row with at least y keystrokes per minute), it's safe to assume that I don't intend to send a shortcut in between.

Do you think, to (optionally) disabling all `tap-hold-next-release` keys while typing quickly is a useful feature? Would it be difficult to implement?

**Comment by thoelze1** [comment]
As long as you think that the benefits of this logic will outweigh its mistakes, I think you should try it. I can imagine edge cases where using a keystrokes-per-minute heuristic would still "misinterpret" keypresses: for example, I finish quickly typing a sentence and immediately attempt to use a shortcut to delete the last word of that sentence. It could certainly fix some instances of problem 2, but could introduce problems of its own. For me, replacing `tap-hold-next` with `tap-hold-next-release` fixed so many instances of problem 2 (home row mods were previously unusable) that it was well worth the (ver few, for me) problems it introduced.

Problem 1 usually happens to me when attempting to use  a hard to reach key, like using `s-TAB` to change windows. I have `super` as part of a `tap-hold-next-release` under my right pinky and `tab` in the default position (left of my left pinky). I see this not as "releasing the first key (under my right pinky) too quickly" but as "releasing the second key (reaching with my left pinky) too late." In my mind, this situation isn't a problem with kmonad or my typing but with the location of `tab` - if I had `tab` in a more accessible location, it wouldn't take me so long to release it.

> Do you think, to (optionally) disabling all tap-hold-next-release keys while typing quickly is a useful feature? Would it be difficult to implement?

I do think it would be a useful feature for people if we could find the right way of measuring whether the user is currently "typing fast." I'm guessing that the better your way of measuring, the more difficult the implementation would be

Another more general solution for determining whether coinciding keypresses are intended to be independent would be to train some kind of model on your keypresses as you use kmonad. If ever you encounter one of the problems you describe, you could tell the model that it misclassified the keypress sequence and (maybe) eventually it would get better over time

**Comment by Rougemoot** [comment]
Using them a lot (with QMK, previously with kmonad), to me home row mods are a matter of:

1) Finding the right timing for your mod-taps.  
This will most likely be different for different keys, as you type faster with some fingers (so it's more sensitive), and some modifiers are never going to be used in a "quick" fashion (so they're less sensitive).  
Finding the right settings can take some time, and you may have to adjust them from time to time because your typing speed will change, but it's a crucial point of using home row mods.

2) Re-learning to use mods.  
This is especially important for shift of course, since it's the one you use the most, and you use it while typing, so it shouldn't interrupt your flow.  
This can be annoying, and it's probably not the answer you're looking for, but I've kept that in mind when I got accidental alphas. To me, setting a bit higher of a tapping term and learning to release the letter before the mod changed the game. It doesn't even really slow you down, after a while.  
Accidental mods while you type **shouldn't** happen, and they are purely a settings problem, though. You shouldn't have to re-learn how to type without accidentally triggering modifiers.

I've been using home row mods for months on my mech now and I never get accidental mods anymore. The occasional accidental alpha, sure, but as you said those are not a problem.

TL;DR: Get your settings to a point where you only get accidental alphas, then try to eliminate those by both tuning your settings and adjusting your muscle memory when you use mods.

Then again I realise this is not really what you're asking for, but IMO the feature you're thinking of is unnecessary and might cause more issues than it solves. But hey, I might be wrong.

**Comment by MaxGyver83** [comment]
> As long as you think that the benefits of this logic will outweigh its mistakes, I think you should try it.

Unfortunately, I can't implement this myself because I don't know Haskell.

> I can imagine edge cases where using a keystrokes-per-minute heuristic would still "misinterpret" keypresses: for example, I finish quickly typing a sentence and immediately attempt to use a shortcut to delete the last word of that sentence.

OK, this could happen. I suppose there is a delay between noticing a typo and using a shortcut for deleting the last word. But maybe there is none. Or none that could be identified as such.

> I do think it would be a useful feature for people if we could find the right way of measuring whether the user is currently "typing fast." I'm guessing that the better your way of measuring, the more difficult the implementation would be

I think there should be a parameter that everybody could adapt to their personal typing speed.

> Another more general solution for determining whether coinciding keypresses are intended to be independent would be to train some kind of model on your keypresses as you use kmonad. If ever you encounter one of the problems you describe, you could tell the model that it misclassified the keypress sequence and (maybe) eventually it would get better over time

This sounds good!

**Comment by Rougemoot** [comment]
No, it also means that you'll get Ctrl-OtherKey if you press OtherKey after 500ms and release your keys in whichever order.
There might be some other settings in QMK, but none that I can think of at the moment. `tap-hold-next-release` acts the same exact way as a mod-tap with `IGNORE_MOD_TAP_INTERRUPT` does in QMK, and that's exactly what you want for home row mods.

**Comment by MaxGyver83** [comment]
OK, thanks. But I guess that I press every key shorter than 500 ms while typing quickly.

Some more detailed logging output could help analyzing what's my typical numbers for `Pe Pn Re Rn` and `Pe Pn Rn Re`. I would need these timestamps:

```
e|CTRL   ‾‾\_____/‾‾‾
n|SHIFT  ‾‾‾‾‾\___/‾‾
           |  |  ||
           e_down
              n_down
                 e_up
                  n_up
```

I already looked into KMonad's source code and tried to figure out how to add these timestamps but I seems to be more complicated than I thought (without adequate Haskell knowledge).

**Comment by david-janssen** [comment]
> Do you think, to (optionally) disabling all tap-hold-next-release keys while typing quickly is a useful feature? Would it be difficult to implement?

At the moment: very. We turn all the button configurations into simple actions, losing the label of what they were specified as. I.e. I have some action that gets triggered by some button, but I don't know if that action was first defined as a tap-next, tap-hold, multi-tap etc.

I'll leave this open and tag it as `someday`, because it's good to keep ideas people have in mind when I think about redesigning the model.

**Comment by klausweiss** [comment]
Hey. Not a Kmonad user (yet), but wanted to give my two cents. I really thought home-row mods can't be comfortable, since you have to embrace a _staccato_ style of typing (relearning to type with [quick swift taps](https://precondition.github.io/home-row-mods#quick-swift-taps)). I had even used them for well over a year, but gave up in the end.

This was until I found out about ZMK's then `global-quick-tap`, now `require-prior-idle-ms` ([docs](https://zmk.dev/docs/behaviors/hold-tap#require-prior-idle-ms)) and a QMK fork which implements it ([blog post](https://sunaku.github.io/home-row-mods.html), [PR](https://sunaku.github.io/home-row-mods.html#patches)). The feature is basically what this very issue is about - disabling HRMs when typing fast. I gave the QMK fork a shot and can confirm it's the bee's knees, am using it ever since. You don't have to _staccato_, but can just type as normal.

Anyway, I wanted to second the feature request and point anyone who wants to try what it feels like to the implementations for ZMK and QMK. Would love to help with the implementation, but I'm afraid I can't spare the time at this point.

Kmonad looks awesome BTW, can't wait to give it a try!

**Comment by Alounv** [comment]
I found something that works quite well in my case.

I simply remove home row mod with a layer when the previous key has been typed less than 200 ms before.

```
    ;; do not use home row mod if previous key has been typed less than 200ms ago
    nmd (layer-delay 200 nmd)
    
    nq #(@q @nmd)
    nw #(@w @nmd)
    ... for all keys ...
```

**Comment by argenkiwi** [comment]
Yet another approach to skipping home row modifiers while typing at speed in Kanata: https://github.com/jtroo/kanata/discussions/1455#discussion-7766783
